### ISSUE TITLE
Tonic and Asymmetric Tonic generate the same results

### ISSUE THREAD

**Issue by thenineteen** [thenineteen/Semiology-Visualisation-Tool#129]
the names of the semiologies are different, but the numbers are exactly the same

![other complex motor](https://user-images.githubusercontent.com/10786099/86423301-711df280-bcd7-11ea-8389-748b4c8efb4b.png)
![asymmetric tonic left](https://user-images.githubusercontent.com/10786099/86423303-71b68900-bcd7-11ea-93cb-56437b742046.png)
![tonic left](https://user-images.githubusercontent.com/10786099/86423306-724f1f80-bcd7-11ea-9580-c339b0d9d92b.png)

**Comment by fepegar** [comment]
This is not related to Slicer:

```python
 Py 3.6.10 (svt)  ~  python -c "import mega_analysis; np = mega_analysis.Semiology('Tonic', symptoms_side=mega_analysis.semiology.Laterality.LEFT, dominant_hemisphere=mega_analysis.semiology.Laterality.LEFT).get_num_datapoints_dict(); print(np)"
Excluded post-ictal semiology in specific query
{32: 3.0, 33: 3.0, 39: 0.0, 40: 1.0, 41: 0.0, 42: 1.0, 48: 3.0, 49: 3.0, 72: 0.0, 73: 0.0, 74: 0.0, 86: 1.0, 94: 1.0, 105: 3.0, 106: 2.0, 107: 1.0, 108: 1.0, 109: 3.0, 110: 3.0, 113: 1.0, 114: 1.0, 115: 3.0, 116: 3.0, 117: 3.0, 118: 3.0, 119: 2.0, 120: 1.0, 121: 2.0, 122: 1.0, 123: 3.0, 124: 3.0, 125: 3.0, 126: 2.0, 129: 3.0, 130: 3.0, 133: 3.0, 134: 3.0, 135: 3.0, 136: 3.0, 137: 3.0, 138: 2.0, 141: 5.0, 142: 4.0, 143: 3.0, 144: 1.0, 145: 3.0, 146: 3.0, 147: 3.0, 148: 2.0, 149: 1.0, 150: 1.0, 151: 2.0, 152: 1.0, 153: 3.0, 154: 1.0, 155: 3.0, 156: 3.0, 157: 3.0, 158: 3.0, 161: 3.0, 162: 3.0, 163: 2.0, 164: 1.0, 165: 2.0, 166: 1.0, 169: 2.0, 170: 2.0, 171: 3.0, 172: 3.0, 175: 1.0, 176: 1.0, 177: 1.0, 178: 1.0, 179: 3.0, 180: 2.0, 181: 3.0, 182: 3.0, 183: 5.0, 184: 4.0, 185: 3.0, 186: 3.0, 187: 2.0, 188: 1.0, 191: 3.0, 192: 1.0, 193: 6.0, 194: 4.0, 195: 1.0, 196: 1.0, 197: 3.0, 198: 3.0, 199: 1.0, 200: 1.0, 201: 3.0, 202: 3.0, 203: 0.0, 204: 0.0, 205: 2.0, 206: 1.0, 207: 3.0, 208: 3.0}
 Py 3.6.10 (svt)  ~  python -c "import mega_analysis; np = mega_analysis.Semiology('Asymmetric Tonic', symptoms_side=mega_analysis.semiology.Laterality.LEFT, dominant_hemisphere=mega_analysis.semiology.Laterality.LEFT).get_num_datapoints_dict(); print(np)"
Excluded post-ictal semiology in specific query
{32: 3.0, 33: 3.0, 39: 0.0, 40: 1.0, 41: 0.0, 42: 1.0, 48: 3.0, 49: 3.0, 72: 0.0, 73: 0.0, 74: 0.0, 86: 1.0, 94: 1.0, 105: 3.0, 106: 2.0, 107: 1.0, 108: 1.0, 109: 3.0, 110: 3.0, 113: 1.0, 114: 1.0, 115: 3.0, 116: 3.0, 117: 3.0, 118: 3.0, 119: 2.0, 120: 1.0, 121: 2.0, 122: 1.0, 123: 3.0, 124: 3.0, 125: 3.0, 126: 2.0, 129: 3.0, 130: 3.0, 133: 3.0, 134: 3.0, 135: 3.0, 136: 3.0, 137: 3.0, 138: 2.0, 141: 5.0, 142: 4.0, 143: 3.0, 144: 1.0, 145: 3.0, 146: 3.0, 147: 3.0, 148: 2.0, 149: 1.0, 150: 1.0, 151: 2.0, 152: 1.0, 153: 3.0, 154: 1.0, 155: 3.0, 156: 3.0, 157: 3.0, 158: 3.0, 161: 3.0, 162: 3.0, 163: 2.0, 164: 1.0, 165: 2.0, 166: 1.0, 169: 2.0, 170: 2.0, 171: 3.0, 172: 3.0, 175: 1.0, 176: 1.0, 177: 1.0, 178: 1.0, 179: 3.0, 180: 2.0, 181: 3.0, 182: 3.0, 183: 5.0, 184: 4.0, 185: 3.0, 186: 3.0, 187: 2.0, 188: 1.0, 191: 3.0, 192: 1.0, 193: 6.0, 194: 4.0, 195: 1.0, 196: 1.0, 197: 3.0, 198: 3.0, 199: 1.0, 200: 1.0, 201: 3.0, 202: 3.0, 203: 0.0, 204: 0.0, 205: 2.0, 206: 1.0, 207: 3.0, 208: 3.0}
```

**Comment by thenineteen** [comment]
after I created this issue I thought I should check for this. Need to investigate if it is


1. data related and ok
2. regex related and broken
3. other bug

**Comment by thenineteen** [comment]
![image](https://user-images.githubusercontent.com/10786099/86499918-9c5f1b00-bd85-11ea-896f-507d20e4f8aa.png)

new asymmetric tonic

**Comment by thenineteen** [comment]
new tonic

![image](https://user-images.githubusercontent.com/10786099/86499997-15f70900-bd86-11ea-90c7-2f250d4f15b4.png)

**Comment by thenineteen** [comment]
various recent commits corrected the regex for tonic which was missing space. 
test added to test this

**Comment by thenineteen** [comment]
now tonic left:

![image](https://user-images.githubusercontent.com/10786099/86501817-8c026c80-bd94-11ea-8ce6-d32d9e4ef7c7.png)

**Comment by thenineteen** [comment]
@fepegar 

this is indicating a bigger problem:

1. when I run the tests how I think they are being run/ should be run, I am getting correct Tonic and Asymmetric Tonic etc.

2. But when I run them using the checkboxes on slicer, Tonic, Asymmetric Tonic and Atonic give very similar results.

3. I checked the loggings and Atonic returns 313 datapoints from QUERY_SEMIOLOGY `Localising Datapoints relevant to query (?i)Atonic: 313.0`
But when I run using 
```
patient = Semiology(
    'Atonic',
    Laterality.NEUTRAL,
    Laterality.NEUTRAL,
)

heatmap = patient.get_num_datapoints_dict()
print('Result:', heatmap)

```

and breakpoint that region, the returned number is 120.0

**Comment by fepegar** [comment]
I'm trying to run this script

```python
from mega_analysis import Semiology, Laterality

atonic = Semiology(
    'Atonic',
    Laterality.NEUTRAL,
    Laterality.NEUTRAL,
)

tonic = Semiology(
    'Tonic',
    Laterality.NEUTRAL,
    Laterality.NEUTRAL,
)

asym_tonic = Semiology(
    'Asymmetric Tonic',
    Laterality.NEUTRAL,
    Laterality.NEUTRAL,
)

print(atonic.get_num_datapoints_dict())
print(tonic.get_num_datapoints_dict())
print(asym_tonic.get_num_datapoints_dict())

```

but I get this error:

```python-traceback
Traceback (most recent call last):
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/arrays/integer.py", line 148, in safe_cast
    return values.astype(dtype, casting="safe", copy=copy)
TypeError: Cannot cast array from dtype('float64') to dtype('int16') according to the rule 'safe'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "alitest.py", line 1, in <module>
    from mega_analysis import Semiology, Laterality
  File "/private/tmp/mega_analysis/__init__.py", line 1, in <module>
    from .semiology import (
  File "/private/tmp/mega_analysis/semiology.py", line 40, in <module>
    excel_data=excel_path)
  File "/private/tmp/mega_analysis/crosstab/mega_analysis/MEGA_ANALYSIS.py", line 157, in MEGA_ANALYSIS
    df['Localising'].astype('Int16', copy=False)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/generic.py", line 5698, in astype
    new_data = self._data.astype(dtype=dtype, copy=copy, errors=errors)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/internals/managers.py", line 582, in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/internals/managers.py", line 442, in apply
    applied = getattr(b, f)(**kwargs)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/internals/blocks.py", line 625, in astype
    values = astype_nansafe(vals1d, dtype, copy=True)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/dtypes/cast.py", line 821, in astype_nansafe
    return dtype.construct_array_type()._from_sequence(arr, dtype=dtype, copy=copy)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/arrays/integer.py", line 354, in _from_sequence
    return integer_array(scalars, dtype=dtype, copy=copy)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/arrays/integer.py", line 135, in integer_array
    values, mask = coerce_to_array(values, dtype=dtype, copy=copy)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/arrays/integer.py", line 249, in coerce_to_array
    values = safe_cast(values, dtype, copy=False)
  File "/usr/local/Caskroom/miniconda/base/envs/ali/lib/python3.6/site-packages/pandas/core/arrays/integer.py", line 156, in safe_cast
    f"cannot safely cast non-equivalent {values.dtype} to {np.dtype(dtype)}"
TypeError: cannot safely cast non-equivalent float64 to int16
```

**Comment by thenineteen** [comment]
weird I do not get that!

ok I'll revert that casting and check again, thanks

**Comment by thenineteen** [comment]
```
    def test_NegativeLookBehind_Regex(self):
        """
        Test the NLB regex used in differentiating Tonic vs Asymmetric Tonic, atonic, generalised tonic, tonic-clonic etc in SemioDict.
        Note also that the SemioDict is the live one unless specifically specified:
            e.g. as an argument to QUERY_SEMIOLOGY(semiology_dict_path=dummy_semiology_dict_path=dummy_semiology_dict_path)
        """
        patient = Semiology('Tonic', Laterality.NEUTRAL, Laterality.NEUTRAL)
        patient.data_frame = self.df
        tonic_result = patient.query_semiology()

        self.assertIs(type(tonic_result), pd.DataFrame)
        assert not tonic_result.empty
        assert(tonic_result['Localising'].sum() == 1+10)
        assert(tonic_result['Lateralising'].sum() == 1)

        # Now test for Atonic
        patient = Semiology('Atonic', Laterality.NEUTRAL, Laterality.NEUTRAL)
        patient.data_frame = self.df
        atonic_result = patient.query_semiology()

        self.assertIs(type(atonic_result), pd.DataFrame)
        assert not atonic_result.empty
        assert(atonic_result['Localising'].sum() == 999)
        assert(atonic_result['Lateralising'].sum() == 2)

        print('\n9 negative lookbehind regex\n')
```


no matter how I change the SemioDict YAML definition of "Atonic", the result never changes for the atonic case (using dummy data) and the localising sum is always 11. This suggests the SemioDict may not be read in properly....

**Comment by thenineteen** [comment]
however, I've tested, and the semiology_dict is being read in properly if determined.

e.g. by  this test

```
def test_parenthesis_and_caps_QUERY_SEMIOLOGY_with_dictionary(self):
        query = QUERY_SEMIOLOGY(
            self.df,
            # see dummy_SemioDict: equivalent to "Aphasia"
            semiology_term='dummy_test_link_aphasia',
            ignore_case=False,
            # in QUERY_SEMIO re.IGNORECASE is used for dictionary anyway
            semiology_dict_path=dummy_semiology_dict_path,
            col1='Reported Semiology',
            col2='Semiology Category',
        )
        assert(query['Localising'].sum() == 14)
        assert(query['Lateralising'].sum() == 6)
        # print('\n4\n')
```


so possibilities:
1. need to explicitly determine the dictionary path for it to look for it (which I thought it did automatically)
2. for some reason the nested Atonic key isn't found

**Comment by thenineteen** [comment]
and I've tests query_semiology and it does find the resources SemioDict (YAML) fine, and when editing the values, it changes the behaviour as expected.

So maybe for some reason Atonic key isn't found!?

**Comment by thenineteen** [comment]
I've also made a local flat SemioDict (and updated QUERY_SEMIOLOGY to remove references to top level key "semiology") and it still doesn't make a difference to atonic.

**Comment by thenineteen** [comment]
I thought it best to keep test class instances of Semiology() separate. But this still doesn't fix the problem.

The problem is:

I can alter the SemioDict and it will alter the inspect_results for any query in the test `test_dummy_data.py` EXCEPT for Atonic. Can't figure it out.

**Comment by thenineteen** [comment]
this was an issue with the order of the regex in matching semiology_term to the key in SemioDict YAML file. 
Bug fixed in 85f7d021991ab926e0ad1a10dff582dd9f4592e3 #132

**Comment by thenineteen** [comment]
for comparison: 

Tonic Left:

![image](https://user-images.githubusercontent.com/10786099/86647536-0bcc4900-bfd8-11ea-89fd-5b5c6c41416c.png)
